include/dali/core/mm/monotonic_resource.h

```diff
@@ -0,0 +1,203 @@
+// Copyright (c) 2020, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_CORE_MM_MONOTONIC_RESOURCE_H_
+#define DALI_CORE_MM_MONOTONIC_RESOURCE_H_
+
+#include <cassert>
+#include <algorithm>
+#include <utility>
+#include "dali/core/mm/memory_resource.h"
+#include "dali/core/mm/detail/util.h"
+#include "dali/core/mm/detail/align.h"
+#include "dali/core/small_vector.h"
+
+namespace dali {
+namespace mm {
+
+/**
+ * @brief Monotonic resource which returns memory from a preallocated buffer.
+ *
+ * Monotonic buffer resources don't require manual deallocation of the returned pointers.
+ * The memory is freed in bulk when the underlying buffer is freed.
+ */
+class monotonic_buffer_resource : public memory_resource {
```

Comment:
I believe this class should have move and copy ctors (might be deleted, but they should be defined)